### ISSUE TITLE
New function for CTC decoding

### ISSUE THREAD

**Issue by songyuc** [pytorch/pytorch#43101]
## 🚀 Feature
New function for decoding the results trained from CTC Loss.

## Motivation

1.  I want a function that can help me decode the results obtained from a model using CTC Loss.
2. There is such a function in PaddlePaddle as [ctc_greedy_decoder](https://www.paddlepaddle.org.cn/documentation/docs/en/api/layers/ctc_greedy_decoder.html).

## Pitch

This op ought to be able to decode sequences  the following steps:
1. Get the indexes of maximum value for each row in input. a.k.a. numpy.argmax(input, axis=0).
2. For each sequence in result of step1, merge repeated tokens between two blanks and delete all blanks.

## Additional context

See  [ctc_greedy_decoder in PaddlePaddle](https://www.paddlepaddle.org.cn/documentation/docs/en/api/layers/ctc_greedy_decoder.html).

**Comment by colesbury** [comment]
@albanD can you comment on this?

**Comment by mruberry** [comment]
Note that TensorFlow (https://www.tensorflow.org/api_docs/python/tf/nn/ctc_greedy_decoder) also has a ctc greedy decoder and Keras accepts a `greedy` argument to its ctc loss (https://www.tensorflow.org/api_docs/python/tf/keras/backend/ctc_decode).

**Comment by songyuc** [comment]
Hi, @mruberry, I implemented one myself with `torch` and `numpy`, as,
Input: tensor, the prediction of words.
Output: list, list of sentences.
Do you think it is appropriate?

**Comment by mruberry** [comment]
Do you have a github repo you can point us to with the implementation, some docs describing it, and an example using it?

**Comment by lightis-tian** [comment]
> Hi, @mruberry, I implemented one myself with `torch` and `numpy`, as,
> Input: tensor, the prediction of words.
> Output: list, list of sentences.
> Do you think it is appropriate?

@songyuc  could you share a repo abut your implements? I'll be vary thankfully.

**Comment by discort** [comment]
@lightis-tian @mruberry 

Here is `numpy` implementation of `ctc_greedy_decoder`
https://gist.github.com/discort/c656925c0b8248fe42c44313a92b3c5e

**Comment by aureliony** [comment]
Since nobody replied, I wrote my own optimized function. It works for tensors of shape `(batch_size, seq_len, n_classes)`.

```python
def ctc_greedy_decode(input: torch.Tensor, blank: int = 0) -> tuple[torch.Tensor]:
    x = input.argmax(2)
    _, indices = x.unique_consecutive(return_inverse=True)
    indices -= indices.min(dim=1, keepdims=True)[0]
    x = torch.full_like(x, blank, device=x.device).scatter_(1, indices, x)
    mask = (x != blank)
    summed_mask = mask.sum(dim=1)
    x = torch.split(x[mask], summed_mask.tolist())
    return x
```
